Disallow different lhist bounds

lenticularis39 opened the pull request:
<!--
Please provide a description of your change below this comment.

Then please complete the checklist.
-->
bpftrace doesn't currently support having linear histograms with different bounds inside a single map (issue #2121). PR #2153 fixes this issue, but at the cost of a significant increase in BPF program size.

This patch adds detecting this unsupported situation and throwing an error into ResourceAnalyzer.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

viktormalik commented on CHANGELOG.md:
```diff
@@ -30,6 +30,9 @@ and this project adheres to
 
 #### Deprecated
 #### Removed
+- Disallow different lhist bounds in a single map
+  - [#2204](https://github.com/iovisor/bpftrace/pull/2204)
+
```
Move this to "Fixed" please. It has never been supported so it shouldn't be in "Removed".

viktormalik commented on src/ast/passes/resource_analyser.cpp:
```diff
@@ -29,16 +30,22 @@ std::string get_literal_string(Expression &expr)
+    out_ << err_.str();
+  }
+
+  return { std::move(resources_), !err_.str().empty() };
```
Since this is returning `bool`, shouldn't the return type be `std::pair<RequiredResources, bool>`? Or, maybe even better, `std::optional<RequiredResources>`?

lenticularis39 replied:
All other analysers (`FieldAnalyser`, `PortabilityAnalyser`, `SemanticAnalyser`) return `int`. In case of `SemanticAnalyser` it's because it returns the number of the pass it failed in, the other ones I assume are `int` was chosen to easily allow multiple error codes in the future if necessary. Either way, wouldn't it be better to keep it consistent here?

viktormalik replied:
Well, consistency is nice but we don't really make use of the error codes anyway. And since `ResourceAnalyser` already returns something different, I wouldn't try too hard here to be similar to other analysers.

lenticularis39 replied:
OK, that makes sense.

fbs replied:
Also the passes are handled by the pass manager nowdays and that has its own error handling (even though its a bit shitty too). So what you do here doesn't matter too much.

viktormalik reviewed (commented):
Would be nice to have a test checking that `lhist` with different bounds fails.

lenticularis39 commented:
Moved the change to Fixed in changelog, changed `ResourceAnalyser` return value type to `std::optional<RequiredResources>` and added a unit test for `ResourceAnalyser` to check for the new behavior.

viktormalik commented on tests/resource_analyser.cpp:
```diff
@@ -0,0 +1,59 @@
+
+using ::testing::_;
+
+void test(BPFtrace &bpftrace, const std::string &input, int expected_result = 0)
```
Shouldn't be:
```suggestion
void test(BPFtrace &bpftrace, const std::string &input, bool expected_result = true)
```

lenticularis39 replied:
Of course, I must have missed that :)

fbs commented:
woops been slow and now it needs a changelog fixup. Can you do that so we can get this in?

lenticularis39 commented:
Changelog fixed.
